### PR TITLE
feat: push manifest list

### PR DESCRIPTION
**Pull request by morlay**
Related: https://github.com/dagger/dagger/issues/1371

Continue https://github.com/dagger/dagger/pull/1467

<del>
Instead of extends `#Push`, adding new task `#PushManifestList` for creating the manifest list

Refs:
* https://docs.docker.com/engine/reference/commandline/manifest/

Not implatemented, just show api changes
</del>

### CODE REVIEW

**Comment by TomChv** [comment]
I'm not sure it's possible to just push manifest list with Buildkit, at least, it wasn't possible when I worked on multi-platform.

You can take a look at https://github.com/dagger/dagger/pull/1467, it's a proof of concept that implement multi-platform in dagger Europa (old version but there shouldn't be much change in go code)

**Comment by morlay** [comment]
@TomChv 

Yes. You're correct. 

Have to use `Control.Build` to commit meta to `Buildkit Worker` for manifests create.

**Comment by morlay** [comment]
@TomChv 

`platform` should be same as `config`, and need to pass from `#Pull / #Dockerfile` to `#Push`

otherwise, the arch-suffix images  created by `docker manifest create` will got wrong manifest list (sub manifest arch will always same as `Plan.platform`)

File: pkg/universe.dagger.io/docker/build.cue [hunk 1]
```diff
@@ -94,7 +95,9 @@ import (
 		"source":     source
 		"auth":       auth
 		"dockerfile": dockerfile
-		"platforms":  platforms
+		if platform != _|_ {
+			"platform": platform
+		}
```
Comment (TomChv) [inline pkg/universe.dagger.io/docker/build.cue:100]: Platform is also optional in `core.#Dockerfile` so you should not need that if statement

File: pkg/universe.dagger.io/docker/pull.cue [hunk 2]
```diff
@@ -20,18 +20,24 @@ import (
...
+
 	_op: core.#Pull & {
 		"source":      source
 		"resolveMode": resolveMode
 		if auth != _|_ {
 			"auth": auth
 		}
+		if platform != _|_ {
+			"platform": platform
+		}
```
Comment (TomChv) [inline pkg/universe.dagger.io/docker/pull.cue:33]: Platform is optional, you do not need that if statement?

File: pkg/universe.dagger.io/docker/push.cue [hunk 3]
```diff
@@ -20,14 +20,26 @@ import (
...
+			input:    image.rootfs
+			config:   image.config
+			platform: image.platform
+		}
+		for _p, _image in images {
+			inputs: "\(_p)": {
+				input:  _image.rootfs
+				config: _image.config
+			}
+		}
```
Comment (TomChv) [inline pkg/universe.dagger.io/docker/push.cue:43]: Disjunction is better than condition: https://docs.dagger.io/1226/coding-style#favor-disjunctions-over-if-conditions

File: plan/task/dockerfile.go [hunk 4]
```diff
@@ -153,9 +153,12 @@ func (t *dockerfileTask) Run(ctx context.Context, pctx *plancontext.Context, s *
 		return nil, fmt.Errorf("failed to unmarshal image config: %w", err)
 	}
 
+	fmt.Println(opts["platform"], "platform")
```
Comment (TomChv) [inline plan/task/dockerfile.go:156]: 👀

**Review by TomChv — COMMENTED** [review]
Thanks for your amazing contribution :rocket:

@aluzzardi @shykes That's a really interesting PR that follows what I did during autumn.

WDYT?

I left minors comments after a quick look, we must take time and put extra attention on this PR so you can wait before applying my comments.

File: pkg/universe.dagger.io/docker/push.cue (hunk 3 above)
Comment (morlay) [reply pkg/universe.dagger.io/docker/push.cue:43]: Like this?
```cue

// Upload an image to a remote repository
#Push: {
	// Destination ref
	dest: #Ref

	// Complete ref after pushing (including digest)
	result: #Ref & _push.result

	// Registry authentication
	auth?: {
		username: string
		secret:   dagger.#Secret
	}

	_push: core.#Push & {
		"dest": dest
		if auth != _|_ {
			"auth": auth
		}
	}

	{
		// Image to push
		image: #Image

		_push: {
			input:    image.rootfs
			config:   image.config
			platform: image.platform
		}
	} | {
		// Images to push
		images: [K=string]: #Image

		_push: inputs: {
			for _p, _image in images {
				"\(_p)": {
					input:  _image.rootfs
					config: _image.config
				}
			}
		}
	}
}
```

File: pkg/dagger.io/dagger/core/image.cue [hunk 5]
```diff
@@ -47,11 +47,24 @@ import (
...
 	// Container image config
-	config: #ImageConfig
+	config?: #ImageConfig
+
+	// Platform of image
+	platform?: string
+
+	{
+		// Filesystem contents to push
+		input: dagger.#FS
```
Comment (TomChv) [inline pkg/dagger.io/dagger/core/image.cue:58]: You could put config in the disjunction too

**Review by TomChv — COMMENTED** [review]
I left another DX comment, I take care of that PR next week

File: pkg/dagger.io/dagger/core/image.cue (hunk 5 above)
Comment (morlay) [reply pkg/dagger.io/dagger/core/image.cue:52]: Yes. `platform` could mv to here.

Could we extract to a `core.#Image` ?

**Review by TomChv — COMMENTED** [review]
I just rebased your PR on main to make sure I have last update, 

I start playing with your PR :rocket:

**EDIT n°1**

 I updated `pull` test to our current coding style, fixed cache test (commit: https://github.com/dagger/dagger/pull/2310/commits/99a39733ad748706c3268cfa52d775654ec8e440) and make sure pull platform is executed in `.bats`

Previously we were doing `dagger do -p pull-platform.cue` but this was not executing anything.

```shell
dagger do -p ./pull_platform.cue     
Usage: 
  dagger do  [flags]

Options


Available Actions:
 test-linux/amd64 
 test-linux/arm64 
 test-linux/s390x 
...
```

It's because now dagger do not run the whole plan, it run a set of defined `actions`.

**EDIT n°2**

I updated push manifest test according to our new standard and same as pull error, tests were not executed.

I love the result

```code
SOPS_AGE_KEY_FILE=./age_key.txt dagger do -p ./tasks/push/push_multi_platform.cue test
[✔] client.commands.sops                                                                                                                                          0.0s
[✔] actions                                                                                                                                                       0.0s
[✔] actions.build."linux/s390x"                                                                                                                                   0.6s
[✔] actions.build."linux/arm64"                                                                                                                                   0.5s
[✔] actions.build."linux/amd64"                                                                                                                                   0.5s
[✔] actions.sopsSecrets                                                                                                                                           0.0s
[✔] actions.push                                                                                                                                                  1.3s
[✔] actions.test."linux/s390x".image                                                                                                                              0.0s
[✔] actions.test."linux/amd64".image                                                                                                                              0.0s
[✔] actions.test."linux/arm64".image                                                                                                                              0.0s
[✔] actions.test."linux/s390x".verify                                                                                                                             0.0s
[✔] actions.test."linux/amd64".verify                                                                                                                             0.0s
[✔] actions.test."linux/arm64".verify                                                                                                                             0.0s
```

**Review by TomChv — COMMENTED** [review]
After some hours of test and inspection, that works really good!
You strongly improved the go code and DX :rocket: 

API changes looks good to me and are not breaking change (which is really good) , but I think some pairs of eyes are required to confirm (@shykes @aluzzardi @helderco @sipsma ).

This PR will need time to to be test before merging it since it's a big one but we are in the good way ;)

Sorry by the way that I committed to your PR, it was the easiest solution to fix test ;)

**Comment by morlay** [comment]
never mind. 
Just hope this feature could land asap. 😂

I want to continue do next big thing, multi-arch builder.

**Comment by sipsma** [comment]
@morlay Appreciate the effort here and agree this is an important feature. I have a couple high-level thoughts + concerns.

In terms of your use case, you mentioned that "QEMU is too slow for compiling huge project" and it seems like your intention after this PR is to add support for distributing builds across multiple Buildkit daemons for each platform (correct me if wrong). I like that idea a lot, but it touches on a lot of issues related to support for multiple Buildkit workers that I think we need to design very carefully.

Even for just this PR, we may need to do a lot more iteration to nail down the cue interface, add a lot more tests to make sure it works seamlessly with the rest of the features (including the `--platform` cli flag), etc.

What I'm getting at is that it might be easier to approach this problem differently to unblock your use case faster. Here's one route I could imagine:
1. Use the existing support for specifying `--platform` at the CLI to build and push images for different platforms. Concurrent executions of `dagger do` for each platform can be used, optionally with each one pointing to a different buildkit worker (via the `BUILDKIT_HOST` env var).
2. To assemble the manifest list, just invoke a cli tool like `docker manifest`, `regclient`, `crane`, etc. from a Dagger action. This could be put in the `universe` action repo for re-usability too.
   * I saw that you tried to use `docker manifest` for this previously. Looking quickly at the [CI output you linked to](https://github.com/octohelm/cuemod/runs/6173820154?check_suite_focus=true#step:6:493) and [associated code](https://github.com/octohelm/cuemod/blob/aa21c6f6f9cfa29fa5d649a56e6a096ef884c2a2/cuepkg/tool/docker.cue#L23), I think the issue may just be that you were not providing access to the docker daemon's socket. There's an example of how to make this work in our `ci.cue` [here](https://github.com/dagger/dagger/blob/a3f005c333da06380692def932b45645fb5636c6/ci.cue#L148-L151). Please let me know if I'm misunderstanding this though as I don't have a ton of context other than the logs you posted.

This is just my personal feeling though. If there's some other reasons the above won't work or you'd rather figure out how to get this support merged into `core`, please let me know. I just want to make sure you are aware that we may need to take our time with these features.

File: plan/task/push.go [hunk 6]
```diff
@@ -66,29 +108,122 @@ func (c *pushTask) Run(ctx context.Context, pctx *plancontext.Context, s *solver
...
-	lg.Debug().Str("dest", dest.String()).Msg("export image")
-	resp, err := s.Export(ctx, st, &img, bk.ExportEntry{
+	log.Ctx(ctx).Debug().Str("dest", dest.String()).Str("platform", platforms.Format(platform)).Msg("export image")
+	return s.Export(ctx, st, &img, bk.ExportEntry{
 		Type: bk.ExporterImage,
 		Attrs: map[string]string{
 			"name": dest.String(),
 			"push": "true",
 		},
 	}, pctx.Platform.Get())
```
Comment (morlay) [inline plan/task/push.go:136]: @TomChv Forgot to replace `pctx.Platform.Get()` here to `platform`

**Comment by morlay** [comment]
@sipsma 

> In terms of your use case, you mentioned that "QEMU is too slow for compiling huge project" and it seems like your intention after this PR is to add support for distributing builds across multiple Buildkit daemons for each platform (correct me if wrong). I like that idea a lot, but it touches on a lot of issues related to support for multiple Buildkit workers that I think we need to design very carefully.

I agree. In fact, `QEMU` already enough for most usages. 
Multi-arch builder only need for project like envoy (I'm maintaining the multi-arch istio, https://github.com/resf/istio, the envoy compiling have to use native host) 

Different from `docker buildx`, dagger split steps in to small actions, so builder of  dagger should support share outputs between multi-arch buildkits, instead merging all arch-matched steps in to matched buildkit,  

> Even for just this PR, we may need to do a lot more iteration to nail down the cue interface, add a lot more tests to make sure it works seamlessly with the rest of the features (including the --platform cli flag), etc.

It could be accept multi platforms, if when it set, it could push with the selected platform.

> To assemble the manifest list, just invoke a cli tool like docker manifest, regclient, crane, etc. from a Dagger action. This could be put in the universe action repo for re-usability too.
I saw that you tried to use docker manifest for this previously. Looking quickly at the [CI output you linked to](https://github.com/octohelm/cuemod/runs/6173820154?check_suite_focus=true#step:6:493) and [associated code](https://github.com/octohelm/cuemod/blob/aa21c6f6f9cfa29fa5d649a56e6a096ef884c2a2/cuepkg/tool/docker.cue#L23), I think the issue may just be that you were not providing access to the docker daemon's socket. There's an example of how to make this work in our ci.cue [here](https://github.com/dagger/dagger/blob/a3f005c333da06380692def932b45645fb5636c6/ci.cue#L148-L151). Please let me know if I'm misunderstanding this though as I don't have a ton of context other than the logs you posted.

In github actions, it could. 
Howerver, in my real-world CI, there is no docker.
Since I migrate to `docker buildx` with buildkits in k8s (mutli-arch cluster, `gitlab-runner` as tigger), I never need to use docker daemon.

**This is why the PR needed, for non-docker environment**

Why mv to dagger?
Just the **composible actions** will be make my team work better.
most developers don't want to care how the image build.
And we have to support amd64 & arm64 both, Which increase the complexity of the Dockerfile.

**Comment by sipsma** [comment]
@morlay 

> In github actions, it could. Howerver, in my real-world CI, there is no docker.

That's alright actually, I was just using `docker manifest` as an example but there are other command line tools that can construct and modify manifest lists without needing docker. Examples I'm aware of are:
* [`regctl`](https://github.com/regclient/regclient/blob/main/docs/regctl.md#manifest-commands)
* [`manifest-tool`](https://github.com/estesp/manifest-tool)

Do you think any of those would work for you? If so, those tools could be wrapped in a reusable action and then put under a `universe.dagger.io` package.

**Comment by morlay** [comment]
@sipsma 

> Do you think any of those would work for you?

Nope. Multi-arch builds is the big feature of `buildkit`, I don't want to use another tool to combine manifest list.

In workflow like (webui + go service): (this should be common for projects)
```mermaid
%%{init:{'theme':'base'}}%%
flowchart TD

web_assets["web assets \n (only need to build on any one arch)"] 
go_build_amd64["go build for amd64"]
go_build_arm64["go build for arm64"]
multi_arch_image["build multi-arch image"]

web_assets 
--> go_build_amd64 & go_build_arm64
--> multi_arch_image
```

Without mult-arch support, we have to use 
```
docker do push --platform=linux/amd64
docker do push --platform=linux/arm64
docker do manyfestlist
```
Another issue with those steps, `web assets` will be re-run for each arch. 

With mult-arch support, we could simple run `dagger do push` to ship all images.
With `platform` declare in `dagger` files, it could clear actions' working platform.

So in another words,
`--platform` may use to filter working-platform-matched actions, instead of force defining all actions' working platform (like `docker buildx` did)

My recently projects, using dagger with mult-arch support (binaries: https://github.com/morlay/dagger/releases/tag/v0.3.0-beta)：
https://github.com/octohelm/cuemod/blob/main/dagger.cue
https://github.com/octohelm/kubepkg/blob/main/dagger.cue
https://github.com/innoai-tech/media-toolkit/blob/main/dagger.cue
Will make pipeline simple:
https://github.com/innoai-tech/media-toolkit/blob/main/.github/workflows/ci.yml#L28-L33

**Comment by sipsma** [comment]
@morlay 

> Nope. Multi-arch builds is the big feature of buildkit, I don't want to use another tool to combine manifest list.

I completely get this and absolutely agree we want to take advantage of buildkit's builtin features in the long-term. The only concern I have right now is that this PR makes a lot of decisions on how we want to approach this problem, which means we are committing to it (or forcing us to make backwards incompatible changes in the future). That's why I thought using a separate tool could be a faster option since it would work already today (even if it has some caveats like you said).

If we can find a way to get this same functionality but only by using experimental command line flags rather than modifying our core actions in cue, that might be another somewhat quicker path (but we'd need sign-off from other team members too of course). I could imagine something like this:
1. We update the `--platform` experimental arg to accept multiple platforms. This results in builds being performed for each platform.
2. If multiple platforms are specified there, the `push` action will push a manifest list instead of just a regular single image manifest

The only problem I'm not sure about is the one you mentioned where your `web assets` step is platform-agnostic and thus shouldn't be repeated for different platforms. Some of the LLB ops in buildkit are platform agnostic (local sources and mergeops come to mind), so if you can provide some more details on how that `web assets` step is implemented, I wonder if we can find a way to get that to work too.

**Comment by morlay** [comment]
@sipsma 

> If multiple platforms are specified there, the push action will push a manifest list instead of just a regular single image manifest

It not change old way, still could use `#Push & { image: "" }` to push single image.
Only use `#Push & { images: {} }` will push manifest list with each platform but tag on the manifest list.

tag on manifest list is better than arch-suffix, it could let use change nothing of k8s manifests to deploy to k8s cluster with any arch node, even to k8s cluster with mulit-arch nodes like me.


> The only problem I'm not sure about is the one you mentioned where your web assets step is platform-agnostic and thus shouldn't be repeated for different platforms. Some of the LLB ops in buildkit are platform agnostic (local sources and mergeops come to mind), so if you can provide some more details on how that web assets step is implemented, I wonder if we can find a way to get that to work too.

```shell
# For each step, it is run on docker container. so platform is the input too.
# go bulild will use the build.web's output for `go:embed` 
$ npm run build.web | go build

#  when dagger do build --platform=amd64             
$ npm run build.web --platform=amd64  | go build --platform=amd64

# when dagger do build --platform=arm64               
$ npm run build.web --platform=arm64  | go build --platform=arm64
# the platform changed, and the cache will be hint.
# `npm run build.web` will pull node for arm64 first, then repeat run install & build.web
```

Like docker did, I prefer declare `platform` in `dagger.cue`, instead of use `--platform` to switch.
```dockerfile
FROM --platform=${BUILD_PLATFORM} node:alpine
```

**Comment by morlay** [comment]
> This PR is stale because it has been open 14 days with no activity. Remove stale label or comment or this will be closed in 7 days.

not stale

**Comment by morlay** [comment]
Hi @TomChv @shykes @aluzzardi @helderco @sipsma
Is there a conclusion？does dagger want to support this feature?

**Comment by sipsma** [comment]
Hi @morlay, we discussed this PR the other day as a team and decided that while we all completely agree this feature is extremely important and we need to support it in the fullness of time, we currently don't have time to give its design+implementation the attention it requires, so it would be best to close it for now. To be clear, that's not to say there's anything wrong with your change, it's just that with any foundational feature like this we need to spend a lot of time validating it and considering as many future possibilities as we can to ensure it will be compatible with everything else we have planned.

The other thing we agreed on as a team is that we truly appreciate all the effort you've put into this and are very sorry that we have to reject the change for now. It's not something we do lightly since we love community contributions and adding new features like this that make Dagger better+easier to use. We just have to balance that against other practical considerations right now.

On our side, we are going to be more careful about being clear upfront on whether PRs are likely to be merged or not so we end up making the best use of everyone's time. Also though, we recommend first running proposed features+designs by Dagger maintainers and getting confirmation that it's the right time to pursue them before diving into the implementation. This is briefly mentioned in our [online contributing docs](https://docs.dagger.io/1227/contributing/#scope-of-pull-requests), but I'll also update our `CONTRIBUTING.md` to make this more clear.

**Comment by morlay** [comment]
> Hi @morlay, we discussed this PR the other day as a team and decided that while we all completely agree this feature is extremely important and we need to support it in the fullness of time, we currently don't have time to give its design+implementation the attention it requires, so it would be best to close it for now. To be clear, that's not to say there's anything wrong with your change, it's just that with any foundational feature like this we need to spend a lot of time validating it and considering as many future possibilities as we can to ensure it will be compatible with everything else we have planned.
> 
> The other thing we agreed on as a team is that we truly appreciate all the effort you've put into this and are very sorry that we have to reject the change for now. It's not something we do lightly since we love community contributions and adding new features like this that make Dagger better+easier to use. We just have to balance that against other practical considerations right now.
> 
> On our side, we are going to be more careful about being clear upfront on whether PRs are likely to be merged or not so we end up making the best use of everyone's time. Also though, we recommend first running proposed features+designs by Dagger maintainers and getting confirmation that it's the right time to pursue them before diving into the implementation. This is briefly mentioned in our [online contributing docs](https://docs.dagger.io/1227/contributing/#scope-of-pull-requests), but I'll also update our `CONTRIBUTING.md` to make this more clear.

Fine. I think I have to use my forked dagger. https://github.com/morlay/dagger/releases/tag/latest

**Pull request closed without merging**